Add a DiscreteEuclideanDomain

crcollins opened the pull request:
This is a very rough pass at adding a DiscreteEuclideanDomain. This is not currently functioning. This PR is mostly to check if I am heading in the right direction. I am very uncertain with the config code, so that is almost a pure guess.

The end goal to be able to create optimization problems like the following:
```
import numpy as np
from dragonfly.exd.domains import DiscreteEuclideanDomain
from dragonfly import maximise_function

size = 10
dim = 3
valid_points = np.random.rand(size, dim)
dom = DiscreteEuclideanDomain(valid_points)
maximise_function(lambda x: np.linalg.norm(x), dom, 10)
```

kirthevasank commented:
Looking at the code, here are my comments:

cp_domain_utils.py: The changes look good. But you might also want to change the function definitions of sample_from_cp_domain_without_constraints and sample_from_config_space to include a type for discrete_euclidean.

domains.py: looks good

cartesian_product_gp.py: L588/L891: I see what you have done here. Using euc allows you to use the same options parsed for euclidean domains. Correct? This is not a bad idea, given a) we don't want to have too many options to set and b) most users will not want to customise a GP model to that extent. However, can you put in a comment at the beginning explaining that the euc_xxx options will also be used by discrete_euclidean domains as well.

cp_ga_optimiser.py (discrete_euclidean_mutation): This looks ok. However, I am worried that taking exponentiated distances in high dimensions might mean that most of the mass will just be concentrated at the same point.

crcollins commented:
cp_domain_utils.py: I did not notice those. I have added them, but in looking around it seems like these get called in several places, and some of these places are passing through values. So, should these options be bubbled up through them? In some places it looks like the nn ones have not been bubbled up, so probably not.

cartesian_product_gp.py: I added a note to the top of the file, not sure if it is really helpful.

cp_ga_optimiser.py: Would it then be better to do like prod_discrete_numeric_exp_mutation and add a linear interpolation with a uniform distribution?


To the example that I have listed above, there seem to be lots of checks preventing this type of problem setup. Should I work my way through all of them adding discrete_euclidean cases to all the checks, or should I be wrapping the domain in a CartesianProductDomain?

kirthevasank commented:
cp_ga_optimiser.py: The mutation operation looks ok now.

cp_domain_utils.py: I see your point. The reason we have a sample type for euclidean and integral spaces is because we want to choose between uniformly random and latin hypercube sampling. We don't have those options for discrete or discrete numeric. For discrete euclidean, it doesn't make sense to do Latin hypercube, so on second thought, I may not be the worst idea to take that argument out.
If you choose to keep it, can you use the following order: euclidean, integral, discrete_euclidean, nn and also add a fidel_space_discrete_euclidean_sample_type to sample_from_config_space.

cartesian_product_gp.py: This is ok for now, but I can expand on this once you commit. Can you please add a TODO for me?

As for the last questions, yes - I think the best way to do this would be to use a CartesianProductDomain (CPD). There are multiple checks, and it just streamlines things if everything is in a CPD. The only exception is for purely EuclideanDomains, where we know we can use a bunch of tricks to make things more efficient.

crcollins commented:
Points noted, I will update this PR when I get some more time.

In the mean time, this is where I am with running it with the settings I am wanting. The script looks like this:
```
from argparse import Namespace
import numpy as np
from dragonfly.exd.domains import DiscreteEuclideanDomain, CartesianProductDomain
from dragonfly import maximise_function

size = 10
dim = 3 
space = np.random.rand(size, dim)
dom = CartesianProductDomain([DiscreteEuclideanDomain(space)])
config = Namespace(
    domain_orderings=Namespace(
        index_ordering=[0],
        dim_ordering=[0],
        kernel_ordering=[0],
    ),  
    domain=dom,
)
maximise_function(lambda x: np.linalg.norm(x), dom, 10, config=config)    
```

I am mostly just making up the config settings as I go. This is the current error that I am at:
```
  File "/home/chris/projects/dragonfly/dragonfly/apis/opt.py", line 178, in maximise_function
    max_capital, is_mf=False, options=options, reporter=reporter)
  File "/home/chris/projects/dragonfly/dragonfly/opt/gp_bandit.py", line 998, in gpb_from_func_caller
    return optimiser.optimise(max_capital)
  File "/home/chris/projects/dragonfly/dragonfly/opt/blackbox_optimiser.py", line 189, in optimise
    return self.run_experiments(max_capital)
  File "/home/chris/projects/dragonfly/dragonfly/exd/exd_core.py", line 535, in run_experiments
    self._main_loop_pre()
  File "/home/chris/projects/dragonfly/dragonfly/opt/gp_bandit.py", line 532, in _main_loop_pre
    self._set_next_gp() # set the next GP
  File "/home/chris/projects/dragonfly/dragonfly/opt/gp_bandit.py", line 307, in _set_next_gp
    self._build_new_gp()
  File "/home/chris/projects/dragonfly/dragonfly/opt/gp_bandit.py", line 410, in _build_new_gp
    gp_fitter.fit_gp_for_gp_bandit(self.options.build_new_model_every)
  File "/home/chris/projects/dragonfly/dragonfly/gp/gp_core.py", line 749, in fit_gp_for_gp_bandit
    ret = self.fit_gp(num_samples, method)
  File "/home/chris/projects/dragonfly/dragonfly/gp/gp_core.py", line 793, in fit_gp
    self._optimise_cts_hps_for_given_dscr_hps(dscr_hps)
  File "/home/chris/projects/dragonfly/dragonfly/gp/gp_core.py", line 579, in _optimise_cts_hps_for_given_dscr_hps
    self.hp_tune_max_evals)
  File "/home/chris/projects/dragonfly/dragonfly/gp/gp_core.py", line 462, in <lambda>
    max_evals)
  File "/home/chris/projects/dragonfly/dragonfly/gp/gp_core.py", line 426, in _direct_wrap
    opt_val, opt_pt, _ = direct_ft_maximise(*args)
  File "/home/chris/projects/dragonfly/dragonfly/utils/oper_utils.py", line 196, in direct_ft_maximise
    *args, **kwargs)
  File "/home/chris/projects/dragonfly/dragonfly/utils/oper_utils.py", line 178, in direct_ft_minimise
    cdata
  File "/home/chris/projects/dragonfly/dragonfly/utils/oper_utils.py", line 151, in _objective_wrap
    return (obj(x), 0)
  File "/home/chris/projects/dragonfly/dragonfly/utils/oper_utils.py", line 194, in <lambda>
    min_obj = lambda x: -obj(x)
  File "/home/chris/projects/dragonfly/dragonfly/gp/gp_core.py", line 577, in <lambda>
    cts_tuning_obj = lambda arg: self._tuning_objective(arg, list(given_dscr_hps))
  File "/home/chris/projects/dragonfly/dragonfly/gp/gp_core.py", line 553, in _tuning_objective
    *args, **kwargs)
  File "/home/chris/projects/dragonfly/dragonfly/gp/gp_core.py", line 537, in build_gp
    gp_cts_hps, gp_dscr_hps, other_gp_params=other_gp_params, *args, **kwargs)
  File "/home/chris/projects/dragonfly/dragonfly/gp/cartesian_product_gp.py", line 387, in _child_build_gp
    self.domain_kernel_ordering, self.domain_kernel_params_for_each_domain)
  File "/home/chris/projects/dragonfly/dragonfly/gp/cartesian_product_gp.py", line 841, in _build_kernel_for_domain
    get_euclidean_integral_gp_kernel_with_scale(kernel_type, 1.0, \
  File "/home/chris/projects/dragonfly/dragonfly/gp/euclidean_gp.py", line 896, in get_euclidean_integral_gp_kernel_with_scale
    raise Exception('Unknown kernel type %s!'%(kernel_type))
Exception: Unknown kernel type 0!
```

Any suggestions before I continue down this path?

kirthevasank commented:
Can you set kernel_ordering to `['']` and see if that works?

One of the things I have not tested in this code is loading a domain directly from a code (instead of from a json file). We will have to write some version of `load_config_file` in [cp_domain_utils.py](https://github.com/dragonfly/dragonfly/blob/master/dragonfly/exd/cp_domain_utils.py) which can take in a Namespace and return what `load_config_file` is returning.

crcollins commented:
Well, I have made some progress with `kernel_ordering=['']`. I took and repurposed the config generated for one of the examples to get the other config values. 
```
size = 10
dim = 3 
space = np.random.rand(size, dim)
print(space)
dom = CartesianProductDomain([DiscreteEuclideanDomain(space)])
config = Namespace(
    domain_orderings=Namespace(
        index_ordering=[[i for i in range(dim)]],
        dim_ordering=[['' for i in range(dim)]],
        kernel_ordering=[''],
        name_ordering=[[str(i) for i in range(dim)]],
        raw_name_ordering=[[str(i) for i in range(dim)]],
    ),  
    domain=dom,
)   
maximise_function(lambda x: np.linalg.norm(x), dom, 10, config=config)
```
It can pull the initial samples just fine, the issue now seems to be when training the GP. Down in `get_euclidean_integral_gp_kernel_with_scale()` gp_cts_hps is empty, so it causes problems. I assume there must be some missing link between the discrete euclidean and euclidean GP config code now.

kirthevasank commented:
I have a few working examples now in examples/supernova/in_code_demo.py and examples/synthetic/hartmann6_4/in_code_demo.py.

Can you write up a similar version for the discrete euclidean case, and see if it works. Otherwise, push your changes, and I will take a stab at it.

crcollins commented:
Ah, I now see what you were doing now. I did not really do anything with the this code today, so it is mostly the same as before.

This code I just committed does not work, but shows the kind of API I was thinking of for this.
What is sort of tricky is, the other definitions of dim don't really match up here. So, that may require some prodding to get it to work right.

So, if you run this as is you should get an error here:
```
  File "dragonfly/dragonfly/exd/cp_domain_utils.py", line 302, in _unpack_vectorised_domain
    assert counter == len(x) # Check if number of variables match
```
I did not dig into exactly what is causing it, but it will come down to how dim(s) should be handled for this domain. In reality, I guess you would consider each discrete euclidean domain dim=1 by default, but here, `x` is one of the rows from the list of valid values. As it is setup right now, dim>1 doesn't work as it is just a flat input. Should this be corrected?

kirthevasank commented:
This example is now running. Can you test if the optimisation is happening correctly?

crcollins commented:
Cool. I ran the same test, and it seems as though it is not taking the dimensionality into account.

For example, in that script, if you print the values that are getting passed to the objective, they are vectors of length 1, and not length 3.

kirthevasank commented:
I think I fixed this with the second commit?

- Kirthevasan
Sent from my mobile. Please excuse brevity.

On Sun, Mar 31, 2019, 11:26 PM Chris Collins <notifications@github.com>
wrote:

> Cool. I ran the same test, and it seems as though it is not taking the
> dimensionality into account.
>
> For example, in that script, if you print the values that are getting
> passed to the objective, they are vectors of length 1, and not length 3.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/dragonfly/dragonfly/pull/21#issuecomment-478422772>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKvgM3KYNC3aZ8ddzbq1O6K6-zQfnPVQks5vcXxTgaJpZM4bbVp7>
> .
>

crcollins commented:
Yeah, did not see the update when I was originally writing that comment. I did some more tests with this, and it seems to work. I have removed the draft status of this PR.

kirthevasank commented:
Ok, let me run a few more tests and merge this.

- Kirthevasan
Sent from my mobile. Please excuse brevity.

On Mon, Apr 1, 2019, 12:30 PM Chris Collins <notifications@github.com>
wrote:

> Yeah, did not see the update when I was originally writing that comment. I
> did some more tests with this, and it seems to work. I have removed the
> draft status of this PR.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dragonfly/dragonfly/pull/21#issuecomment-478650680>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKvgM6hmhnbl4k4XhM67KgyX0dikFl3rks5vcjQrgaJpZM4bbVp7>
> .
>

crcollins commented:
Well, it looks like this PR broke other things from the tests.
https://travis-ci.org/dragonfly/dragonfly/jobs/514342150

kirthevasank commented:
This was something I had not foreseen. I have pushed a fix to it now, in the main branch, but one downside is that you cannot specify a dim parameter for discrete Euclidean domains. In practice, this means that you cannot replicate the same discrete euclidean domain multiple times, short of specifying it multiple times. I don't anticipate many use cases for this, and even if so, it shouldn't take too much effort for a user to do this.
